src/SFML/System/Clock.cpp

```diff
@@ -31,37 +31,102 @@
 
 namespace sf
 {
+
+////////////////////////////////////////////////////////////
+Clock::Clock() noexcept :
+m_refPoint(ClockImpl::now()),
+m_pausePoint()
+{
+
+}
+
+
 ////////////////////////////////////////////////////////////
-Clock::Clock() :
-m_startTime(ClockImpl::now())
+bool Clock::hasTimePassed(const Time& time) const noexcept
 {
+    return time.asMicroseconds() < getElapsedTime().asMicroseconds();
 }
 
 
 ////////////////////////////////////////////////////////////
-Time Clock::getElapsedTime() const
+Time Clock::getElapsedTime() const noexcept
 {
-    return durationToTime(ClockImpl::now() - m_startTime);
+    if (isRunning())
+    {
+        return sf::microseconds((ClockImpl::now() - m_refPoint).count());
+    }
+    else
+    {
+        return sf::microseconds((m_pausePoint - m_refPoint).count());
+    }
 }
 
 
 ////////////////////////////////////////////////////////////
-Time Clock::restart()
+void Clock::start() noexcept
 {
-    const ClockImpl::time_point now = ClockImpl::now();
-    Time elapsed = durationToTime(now - m_startTime);
-    m_startTime = now;
+    m_refPoint += ClockImpl::now() - m_pausePoint;
+
+    m_pausePoint = ClockImpl::time_point{};
+}
+
+
+////////////////////////////////////////////////////////////
+void Clock::pause() noexcept
+{
+    m_pausePoint = ClockImpl::now();
+}
+
+
+////////////////////////////////////////////////////////////
+void Clock::toggle() noexcept
+{
+    if (isRunning())
+    {
+        pause();
+    }
+    else
+    {
+        start();
+    }
+}
 
-    return elapsed;
+
+////////////////////////////////////////////////////////////
+Time Clock::restart() noexcept
+{
+    Time current = getElapsedTime();
+
+    m_pausePoint = ClockImpl::time_point{};
+
+    m_refPoint = ClockImpl::now();
+
+    return current;
+}
+
+
+////////////////////////////////////////////////////////////
+Time Clock::reset() noexcept
+{
+    Time current = getElapsedTime();
+
+    m_pausePoint = m_refPoint = ClockImpl::now();
+
+    return current;
+}
+
+
+////////////////////////////////////////////////////////////
+bool Clock::isRunning() const noexcept
+{
+    return m_pausePoint.time_since_epoch() == ClockImpl::duration::zero();
```

Comment:
Why not `m_pausePoint == ClockImpl::time_point{}`?

Reply:
`m_pausePoint == ClockImpl::time_point()` is also a valid (and maybe simpler) way but I'd personally stick to existing one since it reflects intentions (`m_pausePoint` must be zero to identify that clock is running) better. I'm open to hear others' opinions though.